source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
+# Module system introduction
+
+Note: This tutorial was created from https://github.com/tweag/summer-of-nix-modules, as presented by @infinisil at Summer of Nix 2021, presentation can be seen here (might not work in browser, `mpv` should work): https://infinisil.com/modules.mp4
+
+## Empty module
+
+The simplest module you can have is just an empty attribute set, which
+as you might expect, doesn't do anything!
+
+```diff
+diff --git a/default.nix b/default.nix
+new file mode 100644
+index 0000000..1797133
+--- /dev/null
++++ b/default.nix
+@@ -0,0 +1,3 @@
++{
++  
++}
+```
+
+## Module arguments: lib
+
+Modules can be just an attribute set. But if you want access to some
+arguments you need to change it into a function taking an attribute set
+with an ellipsis (that's the "..."). In this case we only match on the
+`lib` argument, which gives us access to nixpkgs library functions.
+
+Note that the ellipsis is necessary since arbitrary arguments can be
+passed to modules.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 1797133..f7569bd 100644
+--- a/default.nix
++++ b/default.nix
+@@ -1,3 +1,3 @@
+-{
++{ lib, ... }: {
+   
+ }
+```
+
+## Declaring generate.script option
+
+In order for modules to be useful, we need to have options, so let's
+declare one. Options are declared by defining an attribute with
+`lib.mkOption` under the `options` attribute. Here we're defining the
+option `generate.script`.
+
+While there are many attributes to customize options, the most
+important one is `type`, which specifies what values are valid for an
+option, and how/whether multiple values should be merged together.
+
+In the nixpkgs library there are a number of types available under
+`lib.types`. Here we're using the `lines` type, which specifies that:
+- Only strings are valid values
+- Multiple strings are joined with newlines
+
+```diff
+diff --git a/default.nix b/default.nix
+index f7569bd..cb423a9 100644
+--- a/default.nix
++++ b/default.nix
+@@ -1,3 +1,9 @@
+ { lib, ... }: {
++
++  options = {
++    generate.script = lib.mkOption {
++      type = lib.types.lines;
++    };
++  };
+   
+ }
+```
+
+Let's try to evaluate this with this file:
+
+```nix
+# eval.nix
+(import <nixpkgs/lib>).evalModules {
+  modules = [
+    ./default.nix
+  ];
+}
+```
+
+Then we run
+```bash
+nix-instantiate --eval eval.nix -A config.generate.script
+```
+
+Trying to evaluate the `generate.script` option however, we get an error
+that the option is used but not defined, indicating that we need to
```

Comment:
This always confounded me:
- `= config.` _used_: basic¹ user may never see this
- `config. =` _defined_: to his POV this is where he _used_ or _defined_ it
- `options. =` _defined?_: to me, this is where I defined it (I'm not basic user)


_¹basic_:
lower basic nix: never saw a nix file before :baby: 
basic nix modules use: `config. = ""`, use existing options with static values
mid nix modules use: `config. = config.`, use existing options reusing existing values
upper mid nix modules use: `config. = lib.mkIf config.`, use existing options using some kind of logic
higher nix modules use: `options. =`, define his own options

Reply:
In the module system, `config = ` are usually called definitions, while `options = ` are called declarations